processing/src/main/java/org/apache/druid/query/rowsandcols/DecoratedRowsAndColumns.java

```diff
@@ -0,0 +1,246 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.rowsandcols;
+
+import org.apache.druid.frame.Frame;
+import org.apache.druid.frame.FrameType;
+import org.apache.druid.frame.allocation.ArenaMemoryAllocatorFactory;
+import org.apache.druid.frame.key.KeyColumn;
+import org.apache.druid.frame.key.KeyOrder;
+import org.apache.druid.frame.write.FrameWriter;
+import org.apache.druid.frame.write.FrameWriterFactory;
+import org.apache.druid.frame.write.FrameWriters;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.java.util.common.Intervals;
+import org.apache.druid.java.util.common.granularity.Granularities;
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.query.operator.ColumnWithDirection;
+import org.apache.druid.query.rowsandcols.column.Column;
+import org.apache.druid.query.rowsandcols.concrete.FrameRowsAndColumns;
+import org.apache.druid.query.rowsandcols.semantic.DecoratableRowsAndColumns;
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.Cursor;
+import org.apache.druid.segment.StorageAdapter;
+import org.apache.druid.segment.VirtualColumns;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.column.RowSignature;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class DecoratedRowsAndColumns implements DecoratableRowsAndColumns
+{
+  private final RowsAndColumns base;
+
+  private RowsAndColumns materialized = null;
+
+  private Interval interval = null;
+  private Filter filter = null;
+  private VirtualColumns virtualColumns = null;
+  private int limit = -1;
+  private LinkedHashSet<String> viewableColumns = null;
+  private List<ColumnWithDirection> ordering = null;
+
+  public DecoratedRowsAndColumns(
+      RowsAndColumns base
+  )
+  {
+    this.base = base;
+  }
+
+  @Override
+  public Collection<String> getColumnNames()
+  {
+    return viewableColumns == null ? base.getColumnNames() : viewableColumns;
+  }
+
+  @Override
+  public int numRows()
+  {
+    return materializedOrBase().numRows();
+  }
+
+  @Nullable
+  @Override
+  public Column findColumn(String name)
+  {
+    if (viewableColumns != null && !viewableColumns.contains(name)) {
+      return null;
+    }
+
+    return materializedOrBase().findColumn(name);
+  }
+
+  @Nullable
+  @Override
+  public <T> T as(Class<T> clazz)
+  {
+    return null;
+  }
+
+  @Override
+  public void limitTimeRange(Interval interval)
+  {
+    this.interval = interval;
```

Comment:
Throw an error if `interval` was already set? (Otherwise, if it's called twice by mistake, the prior value would be lost.) Unless it's intentional that the prior value be lost if this method is called twice? In which case it would be nice if the javadoc mentioned that.

(Similar comment for the other methods.)